File: apps/robotlab_x_ui/src/models/Peer.ts
// managed
// Peer interface for use across the project
export interface Peer {
  id?: string;
  key?: string;
  url?: string;
  remote_id?: string;
  state?: string;
  upstream_subs?: string[];
  collision?: string;
}

export function createEmptyPeer(): Peer {
  return {
    id: undefined,
    key: undefined,
    url: undefined,
    remote_id: undefined,
    state: undefined,
    upstream_subs: undefined,
    collision: undefined,
  };
}

// Helper function to validate Peer object
export function isPeer(obj: any): obj is Peer {
  return obj && typeof obj === 'object';
}

// Helper function to clone Peer object
export function clonePeer(item: Peer): Peer {
  return {
    ...item,
  };
}

// JSON Schema for Fastify route validation and OpenAPI documentation
export const PeerSchema = {
  "type": "object",
  "properties": {
    "id": {
      "type": "string",
      "description": "Peer key \u2014 the remote_id once known, else the connect URL (primary key)"
    },
    "key": {
      "type": "string",
      "description": "Peer key (same as id)"
    },
    "url": {
      "type": "string",
      "description": "WebSocket URL of the remote runtime",
      "example": "ws://10.0.0.5:8998"
    },
    "remote_id": {
      "type": "string",
      "description": "The peer's runtime id once it identifies; null while still connecting"
    },
    "state": {
      "type": "string",
      "description": "Connection state - identifying, connected, disconnected",
      "example": "connected"
    },
    "upstream_subs": {
      "type": "array",
      "items": {
        "type": "string"
      },
      "description": "Topic names this runtime subscribes to from the peer (bridged upstream)"
    },
    "collision": {
      "type": "string",
      "description": "Set when the peer's runtime id collides with ours; human-readable detail"
    }
  }
}

File: apps/robotlab_x/src/robotlab_x/models/peer.py
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict
from typing import List, Any, Optional, Literal, Dict
import time
import json
import yaml
from database.types import DateTimeStr

class Peer(BaseModel):
    id: Optional[str] = Field(None, description="Peer key — the remote_id once known, else the connect URL (primary key)")
    key: Optional[str] = Field(None, description="Peer key (same as id)")
    url: Optional[str] = Field(None, description="WebSocket URL of the remote runtime", json_schema_extra={"example":"ws://10.0.0.5:8998"})
    remote_id: Optional[str] = Field(None, description="The peer's runtime id once it identifies; null while still connecting")
    state: Optional[str] = Field(None, description="Connection state - identifying, connected, disconnected", json_schema_extra={"example":"connected"})
    upstream_subs: Optional[List[str]] = Field(None, description="Topic names this runtime subscribes to from the peer (bridged upstream)")
    collision: Optional[str] = Field(None, description="Set when the peer's runtime id collides with ours; human-readable detail")

# managed
    @classmethod
    def from_yaml(cls, yaml_str: str):
        """Load Peer from a YAML string."""
        data = yaml.safe_load(yaml_str)
        return cls(**data)

    @classmethod
    def from_yaml_file(cls, file_path: str):
        """Load Peer from a YAML file."""
        with open(file_path, "r") as file:
            return cls.from_yaml(file.read())

    @classmethod
    def from_json(cls, json_str: str):
        """Load Peer from a JSON string."""
        data = json.loads(json_str)
        return cls(**data)

    @classmethod
    def from_json_file(cls, file_path: str):
        """Load Peer from a JSON file."""
        with open(file_path, "r") as file:
            return cls.from_json(file.read())
